File: CodingInterviewsPythonSolutions/2016. Maximum Difference Between Increasing Elements.py
class Solution:
    def maximumDifference(self, nums: List[int]) -> int:
        result = float('-inf')
        n = len(nums)
        maxSoFar = nums[-1]
        for i in range(n - 2, -1, -1):
            if nums[i] < maxSoFar:
                result = max(result, maxSoFar - nums[i])
            maxSoFar = max(maxSoFar, nums[i])
        if result == float('-inf'):
            return -1
        return result

File: CodingInterviewsPackSolutionsCPP/2016. Maximum Difference Between Increasing Elements.cpp
class Solution {
public:
    int maximumDifference(vector<int>& nums) {
        int result = INT_MIN;
        int n = nums.size();
        int maxSoFar = nums.back();
        for(int i = n-2; i >=0; i--){
            if(nums[i]<maxSoFar)
                result = max(result,maxSoFar-nums[i]);

            maxSoFar = max(maxSoFar,nums[i]);
        }
        if(result==INT_MIN)
            return -1;
        return result;
    }
};
